core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -168,45 +169,33 @@ void *impl_allocate_common(const Cuda &exec_space, const char *arg_label,
                            bool exec_space_provided) {
   void *ptr = nullptr;
 
+  const Impl::CudaInternal &cuda_internal =
+      *exec_space.impl_internal_space_instance();
+  cudaError_t error_code;
 #ifndef CUDART_VERSION
 #error CUDART_VERSION undefined!
 #elif (defined(KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC) && CUDART_VERSION >= 11020)
-  cudaError_t error_code;
   if (arg_alloc_size >= memory_threshold_g) {
+    error_code = cuda_internal.cuda_malloc_async_wrapper(&ptr, arg_alloc_size);
+
     if (exec_space_provided) {
-      error_code =
-          exec_space.impl_internal_space_instance()->cuda_malloc_async_wrapper(
-              &ptr, arg_alloc_size);
       exec_space.fence("Kokkos::Cuda: backend fence after async malloc");
     } else {
-      error_code = Impl::CudaInternal::singleton().cuda_malloc_async_wrapper(
-          &ptr, arg_alloc_size);
```

Comment:
I don't see a good reason to use the `singleton` here instead of the provided execution space instance. The default constructed execution space instance corresponds to the singleton anyway.